src/main/java/com/github/maltalex/ineter/range/IPRange.java

```diff
@@ -128,13 +133,29 @@ default boolean contains(IPRange<I, L> range) {
 	 * 
 	 * @return a new range instance
 	 */
-	public IPRange<I, L> withLast(I address);
+	R withLast(I address);
 
 	/**
 	 * Return a new range instance with the <b>same last address</b> as the current
 	 * range, and <b>the given first address</b>
 	 * 
 	 * @return a new range instance
 	 */
-	public IPRange<I, L> withFirst(I address);
+	R withFirst(I address);
+
+	/**
+	 * exclude a range from this range
+	 * @param exclusion the ranges to exclude from original range
+	 * @return A collection containing remaining ranges.
+	 *         If a range is empty - it will not be in the result.
+	 */
+	List<R> withRemoved(Collection<R> exclusion);
```

Comment:
I think with prefix here is confusing since it implies the fluent interface which doesn't work with these methods. I saw a issue suggesting "exclude".

Reply:
Yeah, I'm not a fan of this method name myself, but I couldn't think of a better one. 

What's common to all "with" methods is the creation of a new range (or more in this case), so I felt like "withRemoved" worked. 

I  feel like "exclude" doesn't properly convey the fact that it creates new ranges, and "withExcluded" brings us back to square one.

Other suggestions are more than welcome. Naming things is hard :)

Reply:
I think exclude is fine. I think "withX" is very confusing cause it implies chaining.

Maybe the "collect" prefix is fitting so it implies a collection is returned. Or maybe, inspired by strings, "split"?

As reference:

https://en.m.wikipedia.org/wiki/Fluent_interface

Reply:
First of all, `withRemoved` **can** be used with chaining even if it returns a list. Think about fluent operators like "groupBy" - they can return a list/collection/stream/map of the things being grouped. And even the list itself can be fluently `stream`ed to create meaningful results:

`myRange.withRemoved(otherRange).stream().flatMap(range -> StreamSupport.stream(range.spliterator(),false)).forEach(System.out::println)`

(Btw, I plan on improving ineter's support of `streams` in the near future).

`split` is a bad fit. The resulting list can be empty if you remove the entire range. That's not something a "split" operation should do. And I still don't like exclude because it's a verb, so it feels to me like an action on the original range which isn't what's happening.

I want the name to convey the fact that **new** ranges are being created. How about "without"?

Reply:
I think the fluency is stopped when you return something that is not part of the overall API. If we look at Java streams they do not return collections and collect is the last operation on the streams API. This is because collectors no longer return streams (the grouping collectors return a map as far as I recall).

So in the aspect the operation is not fluent.

What would you like to do with streams?

Reply:
1. While the name `with` can hint at fluency, it doesn't have to. There's a perfectly good explanation why the word "with" is used - it's the common prefix for all methods that return new range instances based on the current one.

2. Fluency isn't stopped when you return something outside the original API. It's ironic you should bring java streams as an example because [this is a thing](https://bugs.openjdk.java.net/browse/JDK-8180352). Also, other fluent APIs like RxJava and LINQ (C#) have terminal operations that return things outside the original API.

3. Despite the above points, I too feel that the current name isn't great. But `exclude` is worse since it implies mutation so I'm sticking with `withRemoved` for now for lack of a better name. Other suggestions are still welcome.

4. Regarding streams - I was thinking about implementing IPRange.stream() using a custom spliterator implementation to allow streaming of individual addressed from a range.